File: core/src/main/java/com/emarsys/core/notification/NotificationManagerProxy.java
package com.emarsys.core.notification;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.os.Build;

import androidx.annotation.RequiresApi;
import androidx.core.app.NotificationManagerCompat;

import com.emarsys.core.util.Assert;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class NotificationManagerProxy {
    private NotificationManager notificationManager;
    private NotificationManagerCompat notificationManagerCompat;

    public NotificationManagerProxy(NotificationManager notificationManager, NotificationManagerCompat notificationManagerCompat) {
        Assert.notNull(notificationManager, "NotificationManager must not be null!");
        Assert.notNull(notificationManagerCompat, "NotificationManagerCompat must not be null!");

        this.notificationManager = notificationManager;
        this.notificationManagerCompat = notificationManagerCompat;
    }

    public boolean areNotificationsEnabled() {
        return notificationManagerCompat.areNotificationsEnabled();
    }

    public int getImportance() {
        return notificationManagerCompat.getImportance();
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public List<ChannelSettings> getNotificationChannels() {
        List<ChannelSettings> channelSettings = new ArrayList<>();

        for (NotificationChannel notificationChannel : notificationManager.getNotificationChannels()) {
            channelSettings.add(new ChannelSettings(notificationChannel.getId(),
                    notificationChannel.getImportance(),
                    notificationChannel.canBypassDnd(),
                    notificationChannel.canShowBadge(),
                    notificationChannel.shouldVibrate(),
                    notificationChannel.shouldShowLights()));
        }
        return channelSettings;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        NotificationManagerProxy that = (NotificationManagerProxy) o;
        return Objects.equals(notificationManager, that.notificationManager) &&
                Objects.equals(notificationManagerCompat, that.notificationManagerCompat) &&
                Objects.equals(areNotificationsEnabled(), that.areNotificationsEnabled());
    }

    @Override
    public int hashCode() {
        return Objects.hash(notificationManager, notificationManagerCompat, areNotificationsEnabled());
    }
}

File: core/src/main/java/com/emarsys/core/notification/NotificationManagerProxy.kt
package com.emarsys.core.notification

import android.app.NotificationManager
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationManagerCompat
import com.emarsys.core.Mockable
import com.emarsys.core.api.notification.ChannelSettings
import java.util.Objects

@Mockable
class NotificationManagerProxy(
    private val notificationManager: NotificationManager,
    private val notificationManagerCompat: NotificationManagerCompat
) {
    val areNotificationsEnabled: Boolean
        get() = notificationManagerCompat.areNotificationsEnabled()

    val importance: Int
        get() = notificationManagerCompat.importance

    @get:RequiresApi(api = Build.VERSION_CODES.O)
    val notificationChannels: List<ChannelSettings>
        get() {
            val channelSettings: MutableList<ChannelSettings> = ArrayList()
            for (notificationChannel in notificationManager.notificationChannels) {
                channelSettings.add(
                    ChannelSettings(
                        notificationChannel.id,
                        notificationChannel.importance,
                        notificationChannel.canBypassDnd(),
                        notificationChannel.canShowBadge(),
                        notificationChannel.shouldVibrate(),
                        notificationChannel.shouldShowLights()
                    )
                )
            }
            return channelSettings
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as NotificationManagerProxy
        return notificationManager == that.notificationManager &&
                notificationManagerCompat == that.notificationManagerCompat &&
                areNotificationsEnabled == that.areNotificationsEnabled
    }

    override fun hashCode(): Int {
        return Objects.hash(
            notificationManager,
            notificationManagerCompat,
            areNotificationsEnabled
        )
    }
}